Replace classes/interfaces of Aura.Intl with our own.

ADmad opened the pull request:
AuraPHP project is practically dead and Aura.Intl package hasn't had any updates since 2017.

<!---

Describe the big picture of your changes here to communicate to the maintainers why we should accept this pull request. If it fixes a bug or resolves a feature request, be sure to link to that issue.

The best way to propose a feature is to open an issue first and discuss your ideas there before implementing them.

Always follow the [contribution guidelines](https://github.com/cakephp/cakephp/blob/master/.github/CONTRIBUTING.md) when submitting a pull request. In particular, make sure existing tests still pass, and add tests for all new behavior. When fixing a bug, you may want to add a test to verify the fix.

-->

ADmad committed:
Replace classes/interfaces of Aura.Intl with our own.

AuraPHP project is practically dead and Aura.Intl package hasn't had any updates since 2017.

ADmad commented:
- Do we want to consider providing class alias for common used classes like `Package` to make upgrade easier? For e.g. [DbMessagesLoader](https://github.com/ADmad/cakephp-i18n/blob/master/src/I18n/DbMessagesLoader.php) of my I18n plugin uses `Aura\Intl\Pacakage`.

- Should I add argument/return typehints that are missing?

ADmad commented on src/I18n/Exception/CannotFormat.php:
```diff
@@ -0,0 +1,25 @@
+ * Redistributions of files must retain the above copyright notice.
+ *
+ * @copyright     Copyright (c) Cake Software Foundation, Inc. (https://cakefoundation.org)
+ * @copyright     Copyright (c) 2017-2020, Aura for PHP
```
Is this copyright tag acceptable? Ref https://github.com/auraphp/Aura.Intl/blob/3.x/LICENSE#L3

garas replied:
```suggestion
 * @copyright     Copyright (c) 2017, Aura for PHP
```
No changes after 2017, so 2020 is inaccurate, especially if range 2017-2020 specified

othercorey replied:
A range isn't necessary for copyright. We can just declare the date it was changed.

garas commented on src/I18n/Exception/CannotFormat.php:
```diff
@@ -0,0 +1,25 @@
+/**
+ * Could not format a message.
+ */
+class CannotFormat extends Exception
```
Exception -> I18nException
CannotFormat ->CannotFormatException
FormatterNotMapped -> FormatterNotMappedException

othercorey commented:
>     * Do we want to consider providing class alias for common used classes like `Package` to make upgrade easier? For e.g. [DbMessagesLoader](https://github.com/ADmad/cakephp-i18n/blob/master/src/I18n/DbMessagesLoader.php) of my I18n plugin uses `Aura\Intl\Pacakage`.

That gets dangerously close to just forking the repository. Do we want to do that?

markstory commented on src/I18n/Exception/Exception.php:
```diff
@@ -0,0 +1,27 @@
+use Cake\Core\Exception\Exception as CoreException;
+
+/**
+ * Could not format a message.
```
Is this comment still accurate?

markstory commented on src/I18n/FormatterLocator.php:
```diff
@@ -0,0 +1,88 @@
+namespace Cake\I18n;
+
+/**
+ * A ServiceLocator implementation for loading and retaining formatter objects.
```
Should this class be marked as internal?

markstory commented on src/I18n/PackageLocator.php:
```diff
@@ -0,0 +1,99 @@
+use Cake\I18n\Exception\Exception;
+
+/**
+ * A ServiceLocator implementation for loading and retaining translator objects.
```
Should this class be marked as internal?

markstory commented on src/I18n/Translator.php:
```diff
@@ -14,27 +14,91 @@
- * Provides missing message behavior for CakePHP internal message formats.
- *
- * @internal
+ * Translator to translate the message.
```
Should this class stay internal? It was internal before.

ADmad commented:
@markstory To begin with I simply made 1:1 copy of Aura's classes and ensure our test suite was passing. But as you pointed out `Translator` and other locator classes could be made internal. They would allow us to remove the locator interfaces too and it doesn't make much sense to have interfaces for internal classes. I could probably also simply `TranslatorRegistry` a bit then and drop classes like `TranslatorFactory` (our other registries themselves have method to create object instances).

ADmad committed:
Fix copyright for files copied from Aura.

ADmad committed:
Rename I18n exception classes.

ADmad commented:
> That gets dangerously close to just forking the repository. Do we want to do that?

Not really, my only aim was to make 4.2 upgrade easy and try to maintain our BC promise for minors. But I guess we could make a small exception in this case.

othercorey commented:
If we accept an Aura.Intl interface somewhere, then I'd say we should provide support. Otherwise, users will probably need to figure out a php 8 solution long-term on their own.

ADmad commented:
> If we accept an Aura.Intl interface somewhere, then I'd say we should provide support

Currently our users doen't really need to use those interfaces anywhere directly. Everything is tucked away under `Translator` which is marked as `@internal`. For e.g. the only Aura.Intl class my plugin requires to use directly is the `Package` class.

ADmad commented:
`TranslatorRegistry` does provide access to the locators but I doubt anyone accesses them to manipulate stuff.

othercorey commented:
Well using a getter is different that accepting an instance. We shouldn't need to alias for that.

ADmad committed:
Mark I18n\Translator as internal class and remove it's interface.

ADmad committed:
Mark I18n\PackageLocator as internal and remove it's interface.

ADmad committed:
Mark I18n\FormatterLocator as internal.

ADmad committed:
Remove I18n\TranslatorFactory classes.

TranslatorRegistry now directly handles creating Translator instances.

ADmad committed:
Fix psalm failures.

ADmad committed:
Add type hints.

ADmad committed:
Avoid generating I18nException

ADmad committed:
Remove unnecessary Closure nesting.

markstory commented:
Should `composer.json` also be updated?

ADmad committed:
Furthur reduce Closure nesting in I18n package.

ADmad committed:
Remove unneeded dependency Aura.Intl.

ADmad commented:
> Should composer.json also be updated?

Done.

I have simplified the internals of `TranslatorRegistry` and reduced Closure nesting.

To avoid having to explicitly packages for each domain-locale pair our `TranslatorRegistry` used loader classes which are used as fallbacks to get package instances. We also fallback domain feature so that if a message id is not found in particular domain name it's fetched from `default` domain. These features let to lost of closure nesting since Aura's `PackageLocator` class didn't allow directly setting `Package` instance, it only accepted callables which returned package instances. 

So I added a new method `setupPackage()` which allows directly setting `Package` instances to our updated package locator. This considerably simplified things.

ADmad committed:
Remove unneeded exception classes.

ADmad committed:
Remove unneeded cache clear call.

ADmad commented:
This is ready for final review.

ADmad commented:
I'll merge this tomorrow if there is no further feedback.

othercorey commented on src/I18n/FormatterLocator.php:
```diff
@@ -0,0 +1,91 @@
+     * @param callable $spec A callable that returns a formatter object.
+     * @return void
+     */
+    public function set(string $name, callable $spec): void
```
Are you still pushing to change callables closures?

ADmad replied:
I am but method will practically never be used. The 2 formatters we have are already setup through the locator's constructor and I don't really see anyone needed to a new formatter type.

That said even using callables which return formatter instance is not necessary. We can just have a formatter name to class name making and the locator and create and return an instance.

othercorey commented on src/I18n/Package.php:
```diff
@@ -0,0 +1,164 @@
+    /**
+     * Message keys and translations in this package.
+     *
+     * @var array
```
```suggestion
     * @var string[]
```

othercorey commented on src/I18n/Package.php:
```diff
@@ -0,0 +1,164 @@
+    /**
+     * Sets the messages for this package.
+     *
+     * @param array $messages The messages for this package.
```
```suggestion
     * @param string[] $messages The messages for this package.
```

othercorey commented on src/I18n/Package.php:
```diff
@@ -0,0 +1,164 @@
+    /**
+     * Adds new messages for this package.
+     *
+     * @param array $messages The messages to add in this package.
```
```suggestion
     * @param string[] $messages The messages to add in this package.
```

othercorey commented on src/I18n/Package.php:
```diff
@@ -0,0 +1,164 @@
+    /**
+     * Gets the messages for this package.
+     *
+     * @return array
```
```suggestion
     * @return string[]
```

othercorey commented on src/I18n/Package.php:
```diff
@@ -0,0 +1,164 @@
+     * Gets the message of the given key for this package.
+     *
+     * @param string $key the key of the message to return
+     * @return mixed The message translation string, or false if not found.
```
```suggestion
     * @return string|false The message translation string, or false if not found.
```

othercorey commented on src/I18n/Translator.php:
```diff
@@ -118,4 +191,14 @@ protected function resolveContext(string $key, array $message, array $vars)
     }
+
+    /**
+     * An object of type Package
```
```suggestion
     * Returns the translator package
```

othercorey commented on src/I18n/TranslatorRegistry.php:
```diff
@@ -72,40 +92,74 @@ class TranslatorRegistry extends TranslatorLocator
+    }
+
+    /**
+     * An object of type PackageLocator
```
```suggestion
     * Returns the translator packages
```

othercorey commented on src/I18n/PackageLocator.php:
```diff
@@ -0,0 +1,124 @@
+     * @param callable|\Cake\I18n\Package $spec A callable that returns a package or Package instance.
+     * @return void
+     */
+    public function set(string $name, string $locale, $spec): void
```
I guess this is required for compatibility for `set` and `setPackage` are confusingly similar.

ADmad replied:
Initially I had thought of adding a separate method for adding `Package` instance so that the argument could be typehinted for both methods but later just updated `set()` to accept both types to keep code DRY and forgot to remove the `setPackage()` method.

What does everyone prefer? Single method on two separate ones? I can update accordingly.

othercorey commented on src/I18n/TranslatorRegistry.php:
```diff
@@ -227,54 +299,26 @@ public function useFallback(bool $enable = true): void
-                return $package;
-            };
+        if ($package->getFallback()) {
+            return $package;
```
Why would the setter return the package being set?

ADmad replied:
The fallback is meant to be set on the package object itself. So if it already has a fallback set nothing has to be done and same object is returned back without any modification.

othercorey replied:
Sure. I just mean it's strange that the setter is returning anything.

ADmad committed:
Update docblocks

Co-authored-by: othercorey <corey.taylor.fl@gmail.com>

ADmad committed:
Simplify I18n\FormatterLocator.

It's now just a map of name to FCQN.

othercorey commented:
Looks ok to me although I am not very familiar with the use of packages. Anything else we can just re-adjust in cakephp 5.
